### ISSUE TITLE
Applying configuration change {"SUPERVISOR_DELTA":"1”} after first boot of a device breaks container downloads

### ISSUE THREAD

**Issue by dtischler** [balena-os/balena-supervisor#1661]
When provisioning devices in normal fashion via balenaCloud (Add device, download OS image, flash to SD Card, boot up) a device will come online, then approximately 14 minutes later the Supervisor makes a change to Delta behavior, and any container downloads that were in progress are started over, wasting the first 14 minutes of bandwidth and downloads in progress.  Here are a pair of logs showing this behavior:

```
12.04.21 13:03:31 (-0700) Supervisor starting
12.04.21 13:03:38 (-0700) Applying configuration change {"SUPERVISOR_POLL_INTERVAL":"900000","SUPERVISOR_DELTA_VERSION":"3"}
12.04.21 13:03:39 (-0700) Applied configuration change {"SUPERVISOR_POLL_INTERVAL":"900000","SUPERVISOR_DELTA_VERSION":"3"}
12.04.21 13:03:40 (-0700) Creating volume 'data'
12.04.21 13:03:40 (-0700) Creating network 'default'
12.04.21 13:03:41 (-0700) Downloading image 'registry2.balena-cloud.com/v2/769c9f80ac93565dcfea4a25558aeb96@sha256:30b995dbc5000e07a929701a8ae315e2853150cffdf7049e51071f83c12d278a'
12.04.21 13:17:33 (-0700) Supervisor starting
12.04.21 13:17:33 (-0700) Applying configuration change {"SUPERVISOR_DELTA":"1"}
12.04.21 13:17:34 (-0700) Applied configuration change {"SUPERVISOR_DELTA":"1"}
12.04.21 13:17:36 (-0700) Downloading image 'registry2.balena-cloud.com/v2/769c9f80ac93565dcfea4a25558aeb96@sha256:30b995dbc5000e07a929701a8ae315e2853150cffdf7049e51071f83c12d278a'
12.04.21 13:30:32 (-0700) Downloaded image 'registry2.balena-cloud.com/v2/769c9f80ac93565dcfea4a25558aeb96@sha256:30b995dbc5000e07a929701a8ae315e2853150cffdf7049e51071f83c12d278a'
12.04.21 13:30:32 (-0700) Installing service 'aai sha256:9b5731bd90cf09c77472808076a458d2f4f0e543dd91031fb64934a65026f698'
12.04.21 13:30:39 (-0700) Installed service 'aai sha256:9b5731bd90cf09c77472808076a458d2f4f0e543dd91031fb64934a65026f698'
12.04.21 13:30:39 (-0700) Starting service 'aai sha256:9b5731bd90cf09c77472808076a458d2f4f0e543dd91031fb64934a65026f698'
12.04.21 13:30:53 (-0700) Started service 'aai sha256:9b5731bd90cf09c77472808076a458d2f4f0e543dd91031fb64934a65026f698’
```

```
12.04.21 16:55:25 (-0700) Supervisor starting
12.04.21 16:55:41 (-0700) Applying configuration change {"SUPERVISOR_POLL_INTERVAL":"900000","SUPERVISOR_DELTA_VERSION":"3"}
12.04.21 16:55:41 (-0700) Applied configuration change {"SUPERVISOR_POLL_INTERVAL":"900000","SUPERVISOR_DELTA_VERSION":"3"}
12.04.21 16:55:44 (-0700) Creating network 'default'
12.04.21 16:55:44 (-0700) Creating volume 'speedtest-data'
12.04.21 16:55:44 (-0700) Creating volume 'settings'
12.04.21 16:55:44 (-0700) Creating volume 'dashboard-data'
12.04.21 16:55:46 (-0700) Downloading image 'registry2.balena-cloud.com/v2/eda20bc938a0d024859a07881f6f1e87@sha256:510cd5821901599a5d60813614edca5241fdbee44c7705af5c3d6ea9633dfe23'
12.04.21 16:55:46 (-0700) Downloading image 'registry2.balena-cloud.com/v2/12be9d2663948ae6ae35ae4cd1ad95ec@sha256:f073f7faaf2149e8887ffafab8f484ab3c8268f6c74f56a0e00c243b2914d941'
12.04.21 16:55:46 (-0700) Downloading image 'registry2.balena-cloud.com/v2/c4ea5a1add59726f695d80da7ed4787a@sha256:3c15b77c3aad7c90a81e6236141195bcbe31348d52d5f1feb3fbf95e5c74491b'
12.04.21 16:55:46 (-0700) Downloading image 'registry2.balena-cloud.com/v2/d57db5c26a838667341a76b940614db7@sha256:73f876e0c3bd02900f829d4884f53fdfffd7098dd572406ba549eed955bf821f'
12.04.21 16:55:47 (-0700) Downloading image 'registry2.balena-cloud.com/v2/c41a555960e84ef4c4b3a7c4a1eeae76@sha256:ec3e1dbb9e198ff04d62c2d0f6eff74cad7438c7d080b89a28428fb96c8479c4'
12.04.21 16:57:35 (-0700) Downloaded image 'registry2.balena-cloud.com/v2/c41a555960e84ef4c4b3a7c4a1eeae76@sha256:ec3e1dbb9e198ff04d62c2d0f6eff74cad7438c7d080b89a28428fb96c8479c4'
12.04.21 17:02:11 (-0700) Failed to download image 'registry2.balena-cloud.com/v2/12be9d2663948ae6ae35ae4cd1ad95ec@sha256:f073f7faaf2149e8887ffafab8f484ab3c8268f6c74f56a0e00c243b2914d941' due to '(HTTP code 404) no such image - no such image: registry2.balena-cloud.com/v2/12be9d2663948ae6ae35ae4cd1ad95ec@sha256:f073f7faaf2149e8887ffafab8f484ab3c8268f6c74f56a0e00c243b2914d941: No such image: registry2.balena-cloud.com/v2/12be9d2663948ae6ae35ae4cd1ad95ec@sha256:f073f7faaf2149e8887ffafab8f484ab3c8268f6c74f56a0e00c243b2914d941 '
12.04.21 17:02:23 (-0700) Failed to download image 'registry2.balena-cloud.com/v2/c4ea5a1add59726f695d80da7ed4787a@sha256:3c15b77c3aad7c90a81e6236141195bcbe31348d52d5f1feb3fbf95e5c74491b' due to '(HTTP code 404) no such image - no such image: registry2.balena-cloud.com/v2/c4ea5a1add59726f695d80da7ed4787a@sha256:3c15b77c3aad7c90a81e6236141195bcbe31348d52d5f1feb3fbf95e5c74491b: No such image: registry2.balena-cloud.com/v2/c4ea5a1add59726f695d80da7ed4787a@sha256:3c15b77c3aad7c90a81e6236141195bcbe31348d52d5f1feb3fbf95e5c74491b '
12.04.21 17:08:52 (-0700) Supervisor starting
12.04.21 17:08:56 (-0700) Applying configuration change {"SUPERVISOR_DELTA":"1"}
12.04.21 17:08:57 (-0700) Applied configuration change {"SUPERVISOR_DELTA":"1"}
12.04.21 17:08:59 (-0700) Installing service 'mqtt sha256:246cd51101f808074ee0d0aeeaf9d634d54fa9aa8cee59f83a9e3213b86f5b4a'
12.04.21 17:08:59 (-0700) Downloading delta for image 'registry2.balena-cloud.com/v2/12be9d2663948ae6ae35ae4cd1ad95ec@sha256:f073f7faaf2149e8887ffafab8f484ab3c8268f6c74f56a0e00c243b2914d941'
12.04.21 17:09:00 (-0700) Downloading delta for image 'registry2.balena-cloud.com/v2/c4ea5a1add59726f695d80da7ed4787a@sha256:3c15b77c3aad7c90a81e6236141195bcbe31348d52d5f1feb3fbf95e5c74491b'
12.04.21 17:09:00 (-0700) Installed service 'mqtt sha256:246cd51101f808074ee0d0aeeaf9d634d54fa9aa8cee59f83a9e3213b86f5b4a'
12.04.21 17:09:00 (-0700) Starting service 'mqtt sha256:246cd51101f808074ee0d0aeeaf9d634d54fa9aa8cee59f83a9e3213b86f5b4a'
12.04.21 17:09:01 (-0700) Downloading delta for image 'registry2.balena-cloud.com/v2/d57db5c26a838667341a76b940614db7@sha256:73f876e0c3bd02900f829d4884f53fdfffd7098dd572406ba549eed955bf821f'
12.04.21 17:09:11 (-0700) Started service 'mqtt sha256:246cd51101f808074ee0d0aeeaf9d634d54fa9aa8cee59f83a9e3213b86f5b4a'
12.04.21 17:09:10 (-0700)  mqtt  1618272550: mosquitto version 1.6.13 starting
12.04.21 17:09:10 (-0700)  mqtt  1618272550: Config loaded from /mosquitto/config/mosquitto.conf.
12.04.21 17:09:10 (-0700)  mqtt  1618272550: Opening ipv4 listen socket on port 1883.
```

**Comment by cywang117** [comment]
This issue is most like resolved with Supervisor v12.6.5 (not yet in production). See #1683, which ensures that Supervisor sends OS version on provision, meaning that Supervisor's first subsequent target state poll will receive `{ SUPERVISOR_DELTA: 1, SUPERVISOR_DELTA_VERSION: 2|3, SUPERVISOR_POLL_INTERVAL: 900000, ... }` in one response from the backend. This ensures that Supervisor applies these crucial delta-related configs with the first `Applying configuration change` log instead of later.

I think that this change is probably crucial enough, and might resolve existing user frictions surrounding delta config settings. What do you about a self-serve update for this version @20k-ultra?

**Comment by 20k-ultra** [comment]
let's confirm this behaviour on a device running say v12.0.0 of the Supervisor then confirm it stops after v12.6.5.

**Comment by cywang117** [comment]
Will do. I'll try with a large release where seeing this behavior might be more likely.

**Comment by cywang117** [comment]
Might need to hold off on this until an OS version with Supervisor v12.6.5+ is released. (The current latest is available, for RPi3 and balenaFin, is BalenaOS 2.77.0+rev1.{prod|dev} which comes with Supervisor v12.5.10.) I'm commenting because I don't know how to change the SV version that installs with a flashed image, and I'm not sure it's even possible 🙃

**Comment by 20k-ultra** [comment]
You could download the OS image and then edit the /etc/supervisor.conf file to point to your new Supervisor version before putting it into a device and booting up @cywang117

**Comment by cywang117** [comment]
Editing the conf file worked! Here are the logs for 2 devices download a ~600Mb release, showing that this issue has been resolved with #1683:

v12.5.10, showing DELTA config application after image has started downloading:
```
Supervisor starting
Applying configuration change {"SUPERVISOR_POLL_INTERVAL":"900000","SUPERVISOR_DELTA_VERSION":"3"}
Applied configuration change {"SUPERVISOR_POLL_INTERVAL":"900000","SUPERVISOR_DELTA_VERSION":"3"}
Creating volume 'resin-data'
Creating network 'default'
Downloading image 'registry2.balena-cloud.com/v2/7c9aed71d3a7744380172a191aff25bc@sha256:64b2f9a23be2195ab931b518f5e5cfd448fbea7c774f5336c9627c8e370c33cc'
Supervisor starting
Downloading image 'registry2.balena-cloud.com/v2/7c9aed71d3a7744380172a191aff25bc@sha256:64b2f9a23be2195ab931b518f5e5cfd448fbea7c774f5336c9627c8e370c33cc'
Downloaded image 'registry2.balena-cloud.com/v2/7c9aed71d3a7744380172a191aff25bc@sha256:64b2f9a23be2195ab931b518f5e5cfd448fbea7c774f5336c9627c8e370c33cc'
Applying configuration change {"SUPERVISOR_DELTA":"1"}
Applied configuration change {"SUPERVISOR_DELTA":"1"}
Installing service 'main sha256:f8ceee161b22606831c7033bab501e560734cac217e462bb1e5a461b197be192'
Installed service 'main sha256:f8ceee161b22606831c7033bab501e560734cac217e462bb1e5a461b197be192'
Starting service 'main sha256:f8ceee161b22606831c7033bab501e560734cac217e462bb1e5a461b197be192'
Started service 'main sha256:f8ceee161b22606831c7033bab501e560734cac217e462bb1e5a461b197be192'
 main  Idling...
```

With these logs for a pre 12.6.5 device you can see that the Supervisor restarts (almost certainly after applying DELTA: 1), which pauses the image download halfway through. This is fine because of resume-able downloads, but when an image layer is in the order of the 100's of mbs or 1+ gb's, gets interrupted while downloading, and has to restart as a result, this can be pretty annoying.

v12.8.0 (fix was in v12.6.5), showing simultaneous application of DELTA & DELTA_VERSION before image ever starts downloading:
```
Supervisor starting
Applying configuration change {"SUPERVISOR_POLL_INTERVAL":"900000","SUPERVISOR_DELTA":"1","SUPERVISOR_DELTA_VERSION":"3"}
Applied configuration change {"SUPERVISOR_POLL_INTERVAL":"900000","SUPERVISOR_DELTA":"1","SUPERVISOR_DELTA_VERSION":"3"}
Creating volume 'resin-data'
Creating network 'default'
Downloading image 'registry2.balena-cloud.com/v2/7c9aed71d3a7744380172a191aff25bc@sha256:64b2f9a23be2195ab931b518f5e5cfd448fbea7c774f5336c9627c8e370c33cc'
Downloaded image 'registry2.balena-cloud.com/v2/7c9aed71d3a7744380172a191aff25bc@sha256:64b2f9a23be2195ab931b518f5e5cfd448fbea7c774f5336c9627c8e370c33cc'
Installing service 'main sha256:f8ceee161b22606831c7033bab501e560734cac217e462bb1e5a461b197be192'
Installed service 'main sha256:f8ceee161b22606831c7033bab501e560734cac217e462bb1e5a461b197be192'
Starting service 'main sha256:f8ceee161b22606831c7033bab501e560734cac217e462bb1e5a461b197be192'
Started service 'main sha256:f8ceee161b22606831c7033bab501e560734cac217e462bb1e5a461b197be192'
 main  Idling...
```

I tested this previously when integration testing #1683, so this is another confirmation.
